lib/nrf_modem_lib/lte_connectivity/lte_connectivity.c

```diff
@@ -64,6 +75,102 @@ void nrf_modem_fault_handler(struct nrf_modem_fault_info *fault_info)
 	fatal_error_notify_and_disconnect();
 }
 
+
+//TODO: Consider reordering these to be below the connectivity timeout and above the reg timeout
+static void become_active()
+{
+	LOG_DBG("Becoming active");
+	net_if_dormant_off(iface_bound);
```

Comment:
@rlubos can confirm, but my recollection is that this should be `net_if_dormant`. @rlubos do you recall the reason why we went with that?

Reply:
This should still be clarified. Reading `dormant` vs. `carrier` state descriptions, I would have selected `carrier` to be used here but there may well be some reasoning which I don't know as I haven't actively used or familiarized myself with the net_if APIs.

Reply:
I thought @rlubos had mentioned to me at some point that `dormant` is more appropriate generally speaking, but looking back on messages I don't see that interaction


In this specific case I actually wonder whether we could use **_both_** -- PDN could update `dormant` and CEREG could update `carrier`

That would not only match roughly with the respective roles of PDN and CEREG state, but might also significantly simplify this PR. I'll play around with it.

@rlubos do you agree this would be an appropriate use of these two functions?

Reply:
Well, the state-tracking simplifications I was hoping for don't really pan out; `lte_connectivity` needs a way to know when to start the connection timeout, and passing off the state tracking problem to `net_if` by using both `carrier` and `dormant` doesn't accomplish that.


I'm gonna defer to @rlubos 's final judgement on whether it should be communicated with `dormant` vs with `carrier`

Reply:
Ah, offloading the problem to `net_if` very nearly works, but creates an unavoidable concurrency problem. Intelligently triggering timeouts requires checking `net_if_is_up` and there is no way to prevent another thread from calling `net_if_down` between that check and the corresponding timeout initiation. 

Explicitly tracking the state has the very strong advantage of avoiding this kind of concurrency trap, so I am tempted to stick with the current approach

I defer to @rlubos on whether we should use `dormant` or `carrier` for that purpose

Reply:
> In this specific case I actually wonder whether we could use **_both_** -- PDN could update `dormant` and CEREG could update `carrier`

When I read the description for `dormant` and `carrier`, I had this option in mind as well. But let's wait for @rlubos comments.

Reply:
`carrier` state represents whether the interface is physically able to transmit packets (so yes, it could be considered L1 related. `dormant` is a bit vague, as conditions for entering/leaving this state can differ between interface types (joining Thread networking or associating with Wi-FI AP). The Linux documentation says:
> Interface is L1 up, but waiting for an external event, f.e. for a protocol to establish.

That said, I think that for LTE connectivity `dormant` **should** be used regardless. But there's no really a problem to use `carrier` as well if it makes sense (I defer to your knowledge about LTE connectivity) and would make the state handling easier. From the application perspective, there's really no difference.

Reply:
The way I understand `carrier` and `dormant` is that the former is more like the physical connection and the latter is logical. Physical connection is kind of clear for different technologies and CEREG status indicates this with the modem.

As @rlubos said, dormant is rather vague but seeing many places referring to Wi-Fi association or authentication, I'd say this would match with LTE registration, or PDN connectivity, which are kind of same as you cannot transmit data without either one of them.

I've browsed https://github.com/zephyrproject-rtos/zephyr/pull/51240, which introduced `carrier` (`NET_IF_LOWER_UP`) and `dormant` (`NET_IF_DORMANT`) states in the commit message as follows:

* Introduce a new interface flag: NET_IF_LOWER_UP, along with
  corresponding helper functions. The flag should be set/cleared on an
  interface by a network driver/L2 to signalize physical readiness of an
  interface to transmit data (for example cable plugged in).

* Introduce a new interface flag: NET_IF_DORMANT, along with
  corresponding helper functions. The flag should be set on an
  iterface when the interface is not ready to transmit application data,
  for example still not joined a Wi-Fi network.


Based on all above, I'd say that LTE cell lost would clearly be `carrier_off` and LTE PDN connectivity matches rather well with `dormant` state.
But as @rlubos said, these do not make difference to applications.

Reply:
> Even if we are CEREG 4 or PDN DEACT, the modem is still capable of transmitting and receiving packets were there someone in range to receive/transmit them, so carrier counts as being up IMO.

If `carrier` is considered up when we've received CEREG 4, then I'm just thinking about these states in a wrong manner. You cannot really transmit/receive anything when CEREG 4 (cell lost) is indicated. If the net_if semantics are such that you should still say that `carrier` is on in that state, then I do agree we shouldn't set carrier on. As said, I do not know the intended semantics of these states but sounds like `carrier` has nothing to do whether we have the real physical connection in place to transmit data instantly. Yes, you could still do a send() when CEREG 4 has been received but data may only go to the modem buffers. It won't reach the destination.

Reply:
> If the net_if semantics are such that you should still say that carrier is on in that state, then I do agree we shouldn't set carrier on

I am at this point pretty certain that is the case, so if we agree on this I will close this thread as resolved

Reply:
I'll leave this completely to you and maybe @rlubos to decide whether you use carrier or dormant states as I'm not an expert on what the meaning of the states are and seems I've made wrong assumptions already. And as was mentioned earlier, from application point of view it doesn't make a difference.